Cosmos 3546 (#207)
vmware-tanzu/graph-framework-for-microservices: docs/getting_started/DMUpgrade.md

|----------------------------------------------|-----------------------------------------------------|:----:|:--------|
| `force`                                      | Force upgrade of a datamodel                        | bool | `false` |
| `datamodel_backward_compatibility_validator` | Flag to enable/disable backward compatibility check | bool | `false` |

Specify each parameter using the `--key=value` argument to `nexus datamodel build`. For example,

```
$ nexus datamodel build --name orgchart  --artifact_repo=<git_repo> --prev_spec_branch=<master>
```

The above command builds the new datamodel against the master branch of the artifactory repo and force will be false.<br>
If prev_spec_branch is not provided then force will be set to true.

## What is backward compatibility check?
It is a check performed to determine the changes performed on dsl during the datamodel upgrade have backward compatibility with existing crds(source of truth).<br>
At build time :<br>
* if "prev-spec-branch" is not provided, force will be set to true and build/crds dir will be the source of truth if present.

> * if "prev-spec-branch" is provided and artifact_repo is provided, force will be set to false and  source of truth will be the crds dir present in prev_spec_branch of artifact_repo provided

Reviewer: pls check the spacing

```
... false and  source ...

> * if "prev-spec-branch" is provided and artifact_repo is not provided, we will check nexus.yaml for artifact_repo and if it is not present we will error and exit <br><br>

Reviewer: If the prev-spec-branch is provided but the artifact_repo is not provided, the nexus.yaml file is checked for the artifact_repo information. If it is not found, an error is thrown, and the process exits.

> At runtime crds existing in the cluster will be the source of truth 

Reviewer: At runtime:

The existing CRDs deployed in the cluster are considered the source of truth for the backward compatibility check.

# Upgrading the Datamodel

## Purpose

> The purpose of Datamodel Upgrade is to make it possible to reinstall Datamodel without having to first uninstall and then reinstall it.

Reviewer 2: ```suggestion
The purpose of Datamodel Upgrade is to make it possible to upgrade Datamodel without having to first uninstall and then reinstall it.
```

| `artifact_repo`    | The git repo url where all current CRDs exist | string | `git_url` |
| `prev-spec-branch` | Branch of current CRDs to compare to new CRDs | string | `master`  |

### Runtime/Install parameters

| Name                                         | Description                                         | Type | Value   |
|----------------------------------------------|-----------------------------------------------------|:----:|:--------|
| `force`                                      | Force upgrade of a datamodel                        | bool | `false` |
| `datamodel_backward_compatibility_validator` | Flag to enable/disable backward compatibility check | bool | `false` |

Specify each parameter using the `--key=value` argument to `nexus datamodel build`. For example,

```
$ nexus datamodel build --name orgchart  --artifact_repo=<git_repo> --prev_spec_branch=<master>
```

The above command builds the new datamodel against the master branch of the artifactory repo and force will be false.<br>
If prev_spec_branch is not provided then force will be set to true.

## What is backward compatibility check?

> The Backward Compatibility Check is a validation process performed during a datamodel upgrade to ensure that the changes made to the DSL are backward compatible with the existing CRDs (Custom Resource Definitions), which serve as the source of truth.

Reviewer 2: ```suggestion
The Backward Compatibility Check is a validation process performed during a datamodel upgrade to ensure that the changes made to the datamodel is backward compatible with the previously released/installed datamodel.
```

<br> The check is performed at build time and runtime

<br>At build time :<br>

> * If "prev-spec-branch" is not provided, force will be set to true and build/crds dir will be the source of truth if present.

Reviewer 2: ```suggestion
* If "prev-spec-branch" is not provided, backward compatibility check will be skipped.
```

> * If "prev-spec-branch" is provided and artifact_repo is provided, force will be set to false and source of truth will be the 'crds' dir present in prev_spec_branch of artifact_repo provided

Reviewer 2: ```suggestion
* If "prev-spec-branch" is provided and artifact_repo is provided, source of truth will be the datamodel artifacts present in prev_spec_branch of artifact_repo provided
```

> * If the prev-spec-branch is provided but the artifact_repo is not provided, the nexus.yaml file is checked for the artifact_repo information. If it is not found, an error is thrown, and the process exits. <br><br>

Reviewer 2: This should be separated into two cases.
* If the prev-spec-branch is provided but the artifact_repo is not provided, the nexus.yaml file is checked for the artifact_repo information, the built datamodel will be compared with datamodel artifacts present in prev_spec_branch of artifact_repo provided.
* If the prev-spec-branch is provided but the artifact_repo is not available, an error is thrown and the backward compatibility check is flagged as failed.

<br>At runtime:<br>

> * CRDs existing in the cluster will be the source of truth 

Reviewer 2: Please fix this per my high level comments.

## When does backward compatibility check fail?
When the change made to the DSL is determined as backward-incompatible (as defined below) then the backward compatibility check in CI will fail.<br>
To enable this we should provide prev-spec-branch and artifact_repo info to set force to false as mentioned above.

## How to perform force upgrade at runtime?

> 1. Ensure there are no CR objects for the dsl node crd that is being upgraded

Reviewer 2: This is not CR's.
Users will need to delete the datamodel objects through REST API or Nexus SDK.

Port forward port 5001 of nexusapp-mgr pod in cosmos-cd namespace
```

> kubectl port-forward -n cosmos-cd nexus-app-mgr-7865f8df58-h7hbj 5018:5001

Reviewer 2: This step is only applicable for Trion (not for Nexus).
So, for Nexus, we should provide the steps to set the force flags via Nexus cli.
For trion, We can capture this detail in a confluence page.

```
2. Invoke the tenant/upgrade POST end point by providing the TenantID and Force info in payload<br>
Example:<br>
HTTP METHOD: POST<br>
URL: http://localhost:5018/tenant/upgrade <br>
Payload:
```
{
    "TenantID": "v4",
    "DatamodelForceInstall": true
}
```
## What is considered a non-breaking change?

#####  1. Add an optional field in the nexus node (field with "omitempty" tag )

<details>
<summary>Show example</summary>

1. Add an optional field called `Location` (field with `omitempty` tag) in the node
    ```
    type Leader struct {
        // Tags "Root" as a node in datamodel graph
        nexus.Node

       Name          string
       Designation   string
       DirectReports Manager `nexus:"children"`
    +  Location      string  `json:"location,omitempty"`
    }

2. Rebuild your datamodel
   ```
   nexus datamodel build --name orgchart
   ```

   Now, the build would succeed

</details>

#####  2. Modify the REST APIs URIs

<details>
<summary>Show example</summary>

> 1. Remove the GET URI from the spec

Reviewer 2: This is Not correct.
Removal of an API should be backward incompatible, even if it is GET

Author: @xmen4xp current behavior is any change to RestAPISpec is backward compatible and that is how expected behavior was documented in UAT test plan (case 12 and 13) https://confluence.eng.vmware.com/display/NSBU/Datamodel+Upgrade+UAT+Plan @rafal-slominskii @PavitraS please update if I am missing anything

Reviewer 3: @xmen4xp IIRC last time when we were talking about this (and testing, as @vipinakavarpu mentioned in UAT plan), we have established that we allow changes in API. We had this idea (as stated with red here https://miro.com/app/board/uXjVOM8wPKU=/?moveToWidget=3458764537971146485&cot=14 ), but conclusion was to allow it. 
If thats not the case know, it has to be changed in compare library, as rn its working like that

    ```shell
   var LeaderRestAPISpec = nexus.RestAPISpec{
    Uris: []nexus.RestURIs{
    -   {
    -       Uri:     "/leader/{root.Leader}",
    -       Methods: nexus.DefaultHTTPMethodsResponses,
    -   },
        {
            Uri:     "/leaders",
            Methods: nexus.HTTPListResponse,
        },
    },
   }

   // nexus-rest-api-gen:LeaderRestAPISpec
   type Leader struct {
      nexus.Node    Name          string
      Designation   string
      DirectReports Manager `nexus:"children"`
      Location      string  `json:"location,omitempty"`
   }

2. Rebuild your datamodel
   ```
   nexus datamodel build --name orgchart 
   ```

   Now, the build would succeed

</details>

#####  3. Modify the REST APIs annotation

<details>
<summary>Show example</summary>

> 1. Modify the `nexus-rest-api-gen` annotation spec from `LeaderRestAPISpec` to `NewLeaderRestAPISpec`

Reviewer 2: This is not modify, but rename.

   ```shell
   var NewLeaderRestAPISpec = nexus.RestAPISpec{
    Uris: []nexus.RestURIs{
        {
            Uri:     "/leaders",
            Methods: nexus.HTTPListResponse,
        },
    },
   }

   // nexus-rest-api-gen:NewLeaderRestAPISpec <==
   type Leader struct {    
      nexus.Node    Name          string
      Designation   string
      DirectReports Manager `nexus:"children"`
      Location      string  `json:"location,omitempty"`
   }

2. Rebuild your datamodel
   ```
   nexus datamodel build --name orgchart 
   ```

   Now, the build would succeed

</details>

## What is considered a backward incompatible change?

> 

Reviewer 2: We also need examples for changes to RestAPISpec that are flagged as backward incompatible.
